components/lib/push-amazon/src/main/java/mozilla/components/lib/push/amazon/AbstractAmazonPushService.kt

```diff
@@ -0,0 +1,81 @@
+/*
+ * This Source Code Form is subject to the terms of the Mozilla Public
+ *  License, v. 2.0. If a copy of the MPL was not distributed with this
+ *  file, You can obtain one at http://mozilla.org/MPL/2.0/.
+ */
+
+package mozilla.components.lib.push.amazon
+
+import android.content.Context
+import android.content.Intent
+import android.os.Bundle
+import com.amazon.device.messaging.ADM
+import com.amazon.device.messaging.ADMMessageHandlerBase
+import mozilla.components.concept.push.EncryptedPushMessage
+import mozilla.components.concept.push.PushError
+import mozilla.components.concept.push.PushProcessor
+import mozilla.components.concept.push.PushService
+import androidx.annotation.VisibleForTesting
+
+/**
+ * An Amazon Cloud Messaging implementation of the [PushService] for Android devices that support Google Play Services.
+ * [ADMMessageHandlerBase] requires a redundant constructor parameter.
+ */
+abstract class AbstractAmazonPushService : ADMMessageHandlerBase("AbstractAmazonPushService"), PushService {
+    private lateinit var adm: ADM
+
+    override fun start(context: Context) {
+        adm = ADM(context)
+        if (adm.registrationId == null) {
+            adm.startRegister()
+        }
+    }
+
+    @VisibleForTesting(otherwise = VisibleForTesting.PROTECTED)
+    public override fun onRegistered(registrationId: String?) {
+        registrationId?.let { PushProcessor.requireInstance.onNewToken(it) }
+    }
+
+    override fun onRegistrationError(errorId: String?) {
+        errorId?.let { PushProcessor.requireInstance.onError(
+                PushError.MalformedMessage("registration failed: $errorId")
+            )
+        }
+    }
+
+    override fun onUnregistered(registrationId: String?) {
+        registrationId?.let {
+            PushProcessor.requireInstance.shutdown()
+        }
+    }
+
+    @VisibleForTesting(otherwise = VisibleForTesting.PROTECTED)
+    public override fun onMessage(intent: Intent?) {
+        intent?.extras?.let {
+            try {
+                val message = EncryptedPushMessage(
+                    channelId = it.getStringWithException("chid"),
+                    body = it.getStringWithException("body"),
+                    encoding = it.getStringWithException("con"),
+                    salt = it.getStringWithException("enc"),
+                    cryptoKey = it.getStringWithException("cryptokey")
+                )
+                PushProcessor.requireInstance.onMessageReceived(message)
+            } catch (e: NoSuchElementException) {
+                PushProcessor.requireInstance.onError(
+                    PushError.MalformedMessage("parsing encrypted message failed: $e")
+                )
+            }
+        }
+    }
+
+    override fun stop() {
+        if (adm.registrationId != null) {
+            adm.startUnregister()
```

Comment:
We should stop the service not unregister it. Can we tell the underlying service to stop running with `stopSelf` and/or any other things that need to be stopped?

Reply:
Works for me